### PR TITLE
refactor(fmgc): decouple flight phase code

### PR DESCRIPTION
**Pull request by globin**
## Summary of Changes
Refactors the Flight Phase handling and seperates it from being to intermingled with other FMGC logic

## Screenshots (if necessary)
n/a code refactoring

## References
n/a code refactoring

## Additional context
Further refactoring to follow, mostly draft to get initial feedback

## Testing instructions
* C/D and in flight spawning
* check for issues with crz altitude and wrong phases (e.g. prog/perf pages)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/shared/src/flightphase.ts [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
-    GoAround,
-    Done,
+    PREFLIGHT,
+    TAKEOFF,
+    CLIMB,
+    CRUISE,
+    DESCENT,
+    APPROACH,
+    GOAROUND,
+    DONE,
```
Comment (Benjozork) [inline src/shared/src/flightphase.ts:9]: **style:** we generally use UpperCamelCase enum member names in the codebase

File: src/fmgc/src/flightphase/Phase.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
+import { FmgcFlightPhase, isAllEngineOn, isAnEngineOn } from '@shared/flightphase';
+import { ConfirmationNode } from '@shared/logic';
+
+export abstract class Phase {
+    // eslint-disable-next-line no-empty-function
+    init(): void { }
+
+    abstract check(time: any): boolean;
```
Comment (Benjozork) [inline src/fmgc/src/flightphase/Phase.ts:8]: **suggestion:** I would rename `check` to `shouldActivateNextPhase`

File: src/fmgc/src/flightphase/FlightPhaseManager.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    checkPhase(_deltaTime: number): void {
+        if (this.phases[this.phase].check(_deltaTime)) {
+            this.changePhase(this.phases[this.phase].nextPhase);
+        }
+    }
+
+    addOnPhaseChanged(cb: (prev: FmgcFlightPhase, next: FmgcFlightPhase) => void): void {
+        this.phaseChangeListeners.push(cb);
```
Comment (Benjozork) [inline src/fmgc/src/flightphase/FlightPhaseManager.ts:47]: **question:** is `addOnPhaseChanged` actually called anywhere ?

**Review by Benjozork — CHANGES_REQUESTED** [review]

File: src/fmgc/src/flightphase/FlightPhaseManager.ts (hunk 3 above)
Comment (globin) [reply src/fmgc/src/flightphase/FlightPhaseManager.ts:47]: https://github.com/flybywiresim/a32nx/pull/6654/files#diff-1e066709d964d45a102d2029d3b294cbffc322231c19cfb38ff109e973155ad1R257

File: src/shared/src/flightphase.ts (hunk 1 above)
Comment (globin) [reply src/shared/src/flightphase.ts:9]: Hah ok I had only seen https://github.com/flybywiresim/a32nx/blob/master/src/shared/src/autopilot.ts before, will change! and also in the other PR where I touch the file above.

**Review by Benjozork — APPROVED** [review]

**Pull request merged**